src/tracer.cpp

```diff
@@ -41,17 +42,36 @@ uint64_t getId() {
   return distribution(TlsRandomNumberGenerator::generator());
 }
 
+std::string reportingHostname() {
+  // This returns the machine name if the DD_TRACE_REPORT_HOSTNAME is set to true.
+  auto report_hostname_opt = std::getenv("DD_TRACE_REPORT_HOSTNAME");
+  if (report_hostname_opt != nullptr && std::strlen(report_hostname_opt) > 0) {
+    if (std::string(report_hostname_opt) == "true") {
+      char buffer[256];
+      if (!::gethostname(buffer, 256)) {
+        return std::string(buffer);
+      }
+    }
+  }
+  return "";
+}
+
 Tracer::Tracer(TracerOptions options, std::shared_ptr<SpanBuffer> buffer, TimeProvider get_time,
                IdProvider get_id, std::shared_ptr<SampleProvider> sampler)
     : opts_(options),
       buffer_(std::move(buffer)),
       get_time_(get_time),
       get_id_(get_id),
-      sampler_(sampler) {}
+      sampler_(sampler),
+      hostname_(reportingHostname()) {}
```

Comment:
Actually this information is a bit redundant and I'm wondering if we can set it only at the root, so the first span we create for every tracing context. Also, can we do it right before sending the trace to our Agent?

Something like:
1. we start a trace, adding spans and all the tags as usual
2. before flushing, we collect the hostname (if the feature is enabled)
3. we attach the hostname to the first span of the trace before enqueueing it in the buffer

This means that for each trace we send only one hostname tag. What do you think?